docs/proposals/timeseries-partitioning-in-compactor.md

```diff
@@ -124,34 +124,27 @@ T1 - Partition 1-2 was created with hash % 2 == 0, and in order to avoid having
+1. Compactor initializes Grouper and Planner.
+2. Compactor retrieves block's meta.json and call Grouper to group blocks for compaction.
 3. Grouper generates compaction plans:
    1. Grouper groups source blocks into unpartitioned groups.
    2. For each unpartitioned group:
       1. Calculates partition number. Partition number indicates how many partitions one unpartitioned group would be partitioned into based on the total size of indices and number of time series from each source blocks in the unpartitioned group.
       2. Assign source blocks into each partition with partition ID (value is in range from 0 to Partition Number - 1). Note that one source block could be used in multiple partitions. So multiple partition ID could be assigned to same source block. Check the examples in [Compaction Partitioning Examples](#compaction-partitioning-examples)
-      3. Gather all partition IDs got assigned to each source blocks.
-      4. For each source block, save compaction plan file into block-storage. Compaction plan file would look like:
-         ```
-         {"partitionIDs": [0, 1, ...]}
-         ```
-4. Grouper returns compaction plans to Thanos Compactor.
-5. Thanos Compactor iterates over each compaction plan (partitioned group). For each iteration, calls Planner to make sure the group is ready for compaction.
+      3. Generates compaction plan that indicates which partition ID each blocks got assigned.
+4. Grouper returns compaction plans to Compactor.
+5. Compactor iterates over each compaction plan (partitioned group). For each iteration, calls Planner to make sure the group is ready for compaction.
 6. Planner collects all compaction plans which are ready for compaction.
    1. For each compaction plan and for each blocks in the plan:
-      1. Retrieve block partition visit marker file from block-storage based on partition ID of the plan.
-      2. If there is no visit marker file or visit marker file is expired and not in completed status, put visit marker file for this block partition in block-storage. The visit marker file would be named like `partition-<partition_id>-visit-mark.json` and its content would look like:
-         ```
-         {"compactorID":"<current_compactor_id>","visitTime":"<date_now>","status":"pending","version":1}
-         ```
-      4. Otherwise, skip this compaction plan. Because at least one block partition is picked up by other compactor.
-7. Return all ready compaction plans to Thanos Compactor.
-8. Thanos Compactor starts compacting partitioned blocks from compaction plans. Once compaction completed, Thanos Compactor would upload deletion marker for the source blocks in the plan. Also, it would update `partition-<partition_id>-visit-mark.json` to have status set to `completed`.
+      1. Make sure the source block fits within the time range of the group.
+      2. Make sure the source block with assigned partition ID is currently not used by another ongoing compaction.
```

Comment:
How do we make sure this?

Reply:
By checking visit marker file, I removed implementation detail here. But basically we can expand the visit marker file to not only used for entire block, instead there will be one marker file per partition ID per block. So any ongoing compaction on a compactor would claim that partition ID is under use.